server/src-lib/Data/Time/Clock/Units.hs

```diff
@@ -148,24 +371,70 @@ instance (KnownNat picosPerUnit) => RealFrac (TimeUnit picosPerUnit) where
   floor = floor . toRational
 
 -- we can ignore unit:
-instance Hashable (TimeUnit a) where
-  hashWithSalt salt (TimeUnit dt) = hashWithSalt salt $ 
+instance Hashable (TimeUnit a 'Absolute) where
+  hashWithSalt salt (TimeUnit dt) = hashWithSalt salt $
     (realToFrac :: DiffTime -> Double) dt
 
 
+-- instances for Calendar
+-- NOTE: secondsToNominalDiffTime :: Pico -> NominalDiffTime
+-- http://hackage.haskell.org/package/time-1.9.3/docs/Data-Time-Clock.html
+-- If a Pico value is supplied to the function then the output is also in Pico
+-- and when the Pico value is converted into seconds(divide by 10^12)
+-- then the output is in Seconds
+-- So, the function `toFixedPico` converts Pico into Seconds
+instance (KnownNat picosPerUnit) => Num (TimeUnit picosPerUnit 'Calendar) where
+  TimeUnit a + TimeUnit b = TimeUnit $ a + b
+  TimeUnit a - TimeUnit b = TimeUnit $ a - b
+  TimeUnit a * TimeUnit b = TimeUnit . secondsToNominalDiffTime . toFixedPico $
+    (toRational $ nominalDiffTimeToSeconds a * nominalDiffTimeToSeconds b) `div'` (natNum @picosPerUnit)
+  negate (TimeUnit a) = TimeUnit $ negate a
+  abs (TimeUnit a) = TimeUnit $ abs a
+  signum (TimeUnit a) = TimeUnit $ signum a
+  fromInteger a = TimeUnit . secondsToNominalDiffTime . toFixedPico $ a * natNum @picosPerUnit
+
+instance (KnownNat picosPerUnit) => Read (TimeUnit picosPerUnit 'Calendar) where
+  readsPrec _ = map (first fromRational) . readFloat
+
+instance (KnownNat picosPerUnit) => Fractional (TimeUnit picosPerUnit 'Calendar) where
+  TimeUnit a / TimeUnit b = TimeUnit . secondsToNominalDiffTime . toFixedPico $ fromInteger $
+    ((toRational $ nominalDiffTimeToSeconds a) * natNum @picosPerUnit) `div'` (toRational $ nominalDiffTimeToSeconds b)
+  fromRational a = TimeUnit . secondsToNominalDiffTime . toFixedPico $ round (a * natNum @picosPerUnit)
+
+instance (KnownNat picosPerUnit) => Real (TimeUnit picosPerUnit 'Calendar) where
+  toRational (TimeUnit a) = toRational (nominalDiffTimeToSeconds a) / natNum @picosPerUnit
+
+instance (KnownNat picosPerUnit) => RealFrac (TimeUnit picosPerUnit 'Calendar) where
+  properFraction a = (i, a - fromIntegral i)
+    where i = truncate a
+  truncate = truncate . toRational
+  round = round . toRational
+  ceiling = ceiling . toRational
+  floor = floor . toRational
+
+-- we can ignore unit:
+instance Hashable (TimeUnit a 'Calendar) where
+  hashWithSalt salt (TimeUnit dt) = hashWithSalt salt $
+    (realToFrac :: NominalDiffTime -> Double) dt
+
 -- | Duration types isomorphic to 'DiffTime', powering 'fromUnits'.
-class Duration d where
+class Duration' d where
   fromDiffTime :: DiffTime -> d
   toDiffTime :: d -> DiffTime
 
-instance Duration DiffTime where
+instance Duration' DiffTime where
   fromDiffTime = id
   toDiffTime = id
 
-instance Duration NominalDiffTime where
+instance Duration' NominalDiffTime where
```

Comment:
This seems quite sketchy to me—the purpose of this class is to make `fromUnits` work, which should really only be able to convert between units *of the same underlying type*. You should be able to convert from a `Second 'Absolute` to a `Minutes 'Calendar`.

Consider restructuring the `AsDuration` class to look like this:

```haskell
class AsDuration t d | d -> t where
  fromDuration :: Duration t -> d
  toDuration :: d -> Duration t
```

Instances of this class will look like this:

```haskell
instance AsDuration 'Absolute DiffTime where
  fromDuration = id
  toDuration = id

instance AsDuration 'Calendar NominalDiffTime where
  fromDuration = id
  toDuration = id

deriving instance AsDuration t (Duration t) => AsDuration t (Seconds t)
deriving instance AsDuration t (Duration t) => AsDuration t (Minutes t)
-- etc.
```

Now `fromUnits` can be made safer:

```haskell
fromUnits :: (AsDuration t x, AsDuration t y) => x -> y
```

The two uses of `t` in the type signature ensure they are the same.

Reply:
I think the purpose of `fromUnits` is to convert between `NominalDiffTime` and `DiffTime`, 

```
  , fromUnits
  -- * Reexports
  -- | We use 'DiffTime' as the standard type for unit-agnostic duration in our
  -- code. You'll need to convert to a 'NominalDiffTime'  (with 'fromUnits') in
  -- order to do anything useful with 'UTCTime' with these durations.
``` 
I found this note in the `Data.Time.Clock.Units` module. 

The `sleep` function defined in the `Control.Concurrent.Extended` module takes a `DiffTime` as an argument.

I thought of making a new class `Sleep` which will have instances for both `Duration 'Absolute` and `Duration 'Calendar`, but I'm not sure if that's such a good idea.

Let me know what you think.

Reply:
I think it’s unlikely that we want to use `sleep` with anything other than `Duration 'Absolute`. I’m trying to imagine a use case where I would want/need to use `sleep` with `Duration 'Calendar`, but I can’t think of any.

The purpose of distinguishing these types is primarily conceptual. The idea of `Duration 'Absolute` is that it refers to an absolute time duration, which is always the same length. One hour is always one hour. This is what makes sense to pass to `sleep`, since `sleep` tells the processor to yield for a certain, absolute duration of time.

`Duration 'Calendar`, on the other hand, describes an interval between two *calendar datetimes.* What’s interesting is that this corresponds to different absolute times depending on where in the calendar it occurs. For example, a duration of “1 month” might be 744 hours in January, but 672 hours in February… except in leap years, when it’s 696 hours.

Keeping these intervals distinct from absolute intervals is helpful, since they’re really two different ideas. An absolute duration has a universal, physical meaning, but a calendar duration only means something in the context of human society.

Notably, the only way to meaningfully convert between a `Duration 'Absolute` and a `Duration 'Calendar` is to have a calendar datetime as your starting point. If you have a start date, like “January 12th, 13:44 UTC,” and you have a calendar duration like “1 month,” you can combine the two to get an absolute duration of 744 hours. But if you don’t have a calendar datetime to use as your starting point, you can’t *get* an absolute time out.

---

Now, to be fair, worrying about all this in practice doesn’t always make very much sense. The actual `NominalDiffTime` type can’t handle durations like “1 month” because it only stores a certain number of seconds. That means that the only real difference between `DiffTime` and `NominalDiffTime` is that they handle leap seconds differently, and we don’t care if an event is triggered 1 second too early or too late. Leap seconds are so rare that it’s totally irrelevant.

But distinguishing the types is still useful for a programmer, since it means we know which type we’re talking about at any given time, and it makes us be a little more careful to not treat them the same way. If we ever *do* need to handle durations like “1 month,” we can expand/adapt `Duration 'Calendar` to handle it (like by adding in a `CalendarDiffDays` component), with minimal changes to the code.

I think your question might be something of [an XY problem](https://meta.stackexchange.com/a/66378/188673). I assume that, given you’re worried about using `sleep` with a `Duration 'Calendar`, you have a situation in the code where you want to do that. Depending on the details, it might make sense to either:

  1. Just use `Duration 'Absolute` instead. Maybe it isn’t a big deal, I don’t know.

  2. Provide a function to go from a `Duration 'Calendar` to a `Duration 'Absolute` with a suitably scary name like `toAbsoluteIgnoringLeapSeconds` or something like that.

But I think it makes sense to keep `sleep` accepting a `Duration 'Absolute`, since `sleep` shouldn’t decide how to do the conversion, `sleep`’s caller should.

Reply:
I'll keep `sleep` as it is and I agree that `sleep`'s caller should take care of the conversion. 

Thanks for the explanation of `NominalDIffTime`, things make a lot of sense now. 🙂